load-test-framework/src/main/java/com/google/pubsub/flic/controllers/GCEController.java

```diff
@@ -0,0 +1,446 @@
+import com.google.api.services.compute.model.*;
+import com.google.api.services.storage.Storage;
+import com.google.api.services.storage.model.Bucket;
+import com.google.cloud.pubsub.PubSub;
+import com.google.cloud.pubsub.PubSubOptions;
+import com.google.cloud.pubsub.SubscriptionInfo;
+import com.google.cloud.pubsub.TopicInfo;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.SettableFuture;
+import com.google.pubsub.flic.controllers.Client.ClientType;
+import org.apache.commons.codec.digest.DigestUtils;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.util.*;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class GCEController extends Controller {
+  private static final String machineType = "n1-standard-4"; // quad core machines
+  private static final String sourceFamily = "projects/ubuntu-os-cloud/global/images/ubuntu-1604-xenial-v20160930";
+  private final Storage storage;
+  private final Compute compute;
+  private final String projectName;
+  private final Map<String, Map<ClientParams, Integer>> types;
+  private boolean shutdown;
+
+  private GCEController(String projectName, Map<String, Map<ClientParams, Integer>> types,
+                        ScheduledExecutorService executor, Storage storage, Compute compute, PubSub pubSub) throws Throwable {
+    super(executor);
+    this.shutdown = false;
+    this.projectName = projectName;
+    this.types = types;
+    this.storage = storage;
+    this.compute = compute;
+
+    List<SettableFuture<Void>> pubsubFutures = new ArrayList<>();
+    types.values().forEach((paramsMap) -> paramsMap.keySet().stream().map((params) -> params.clientType)
+        .distinct().forEach((clientType) -> {
+          SettableFuture<Void> pubsubFuture = SettableFuture.create();
+          pubsubFutures.add(pubsubFuture);
+          executor.execute(() -> {
+            // Delete each topic and subscription if it exists and create it new to avoid potential backlog from previous runs
+            String topic = Client.topicPrefix + Client.getTopicSuffix(clientType);
+            try {
+              pubSub.create(TopicInfo.of(topic));
+            } catch (Exception e) {
+              log.info("Topic already exists, reusing.");
+            }
+            paramsMap.keySet().stream().filter((params) -> params.clientType == clientType && params.subscription != null)
+                .map((params) -> params.subscription).forEach((subscription) -> {
+              pubSub.deleteSubscription(subscription);
+              pubSub.create(SubscriptionInfo.of(topic, subscription));
```

Comment:
So we create the pubsub resources regardless of whether the load will be generated against pubsub or kafka, right? Do we need to create kafka resources up front as well then?